main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Differ.scala

```diff
@@ -26,15 +26,55 @@ object Differ {
     case None =>
       // Case 1: No old AST. No difference.
       DeltaContext(Nil)
-    case Some(oldAst) => {
+    case Some(oldAst) =>
       // Case 2: We have an oldAst and a newAst. Compute their difference.
 
       // TODO: Add some comments and introduce helper functions.
-      val newDefs = (newAst.defs.keySet -- oldAst.defs.keySet).toList.map(sym => Delta.AddDef(sym, getCurrentTimestamp))
+      val newDefs = calcDefDiff(oldAst, newAst)
       val newEnums = (newAst.enums.keySet -- oldAst.enums.keySet).toList.map(sym => Delta.AddEnum(sym, getCurrentTimestamp))
 
       DeltaContext(newDefs ++ newEnums)
-    }
+  }
+
+  /**
+    * Computes the difference between oldAst and newAst for defs
+    *
+    * @param oldAst the old AST.
+    * @param newAst the new AST.
+    * @return a list of changes as Delta's.
+    */
+  private def calcDefDiff(oldAst: TypedAst.Root, newAst: TypedAst.Root): List[Delta] = {
+    newAst.defs.flatMap {
+      case (sym, newDef) =>
+        // Check if the sym also exists in the oldAst
+        oldAst.defs.get(sym) match {
+          case None =>
+            // It's a newly added def, add it to the deltas
+            Some(Delta.ModifiedDef(sym, getCurrentTimestamp))
+          case Some(oldDef) =>
+            // We have an old def, check if the implementation has changed
+            if (compareDefs(oldDef, newDef)) {
+              // The def has changed, add it to the deltas
+              Some(Delta.ModifiedDef(sym, getCurrentTimestamp))
+            } else {
+              // The def hasn't changed, don't add it to the deltas.
+              None
+            }
+        }
+    }.toList
+  }
+
+  /**
+    * Compares two defs in terms of change
+    *
+    * @param oldDef the old def.
+    * @param newDef the new def.
+    * @return true if the def has changed, false otherwise.
+    */
+  private def compareDefs(oldDef: TypedAst.Def, newDef: TypedAst.Def): Boolean = {
```

Comment:
Why not define the function on two source locations, e.g. `def equalLocations(loc1: ..., loc2: ...): Boolean`. Then you can reuse it for different things.

Reply:
Are you sure that we do not already have the required functionality:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala#L123

Reply:
When I tested it, the source makes all defs appear as recently modified